realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2014 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.dynamic;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import io.realm.Realm;
+import io.realm.RealmObject;
+import io.realm.internal.ColumnType;
+import io.realm.internal.Row;
+import io.realm.internal.Table;
+import io.realm.internal.TableOrView;
+
+/**
+ * Object for interacting with a RealmObject using dynamic names.
+ *
+ * @see io.realm.RealmMigration
+ */
+public class DynamicRealmObject {
+
+     Realm realm;
+     Row row;
+
+    /**
+     * Creates a dynamic Realm object based on a row entry.
+     */
+    public DynamicRealmObject(Realm realm, Row row) {
```

Comment:
I agree, exposing internal classes is bugging me as well. I considered a few other options but neither seem that great either:

1) Moving DynamicRealmObject/DynamicRealmList to ```io.realm``` instead of ```io.realm.dynamic``` -> I would prefer to keep the main package space as clean as possible, i.e. only primary classes/interfaces.

2) Extending RealmObject doesn't work if we keep these classes in ```.dynamic``` as they cannot access the package protected fields from RealmObject.

Generally the current approach where we rely package protected methods/fields to avoid exposing internals is hurting our ability to modularise the code base. I don't have any golden answer though. But we should probably give this some more thought...